tools/build_defs/repo/git_worker.bzl

```diff
@@ -0,0 +1,127 @@
+_GitRepo = provider(
+    doc = "Provider to organize precomputed arguments for calling git.",
+    fields = {
+        "directory": "Working directory path",
+        "shallow": "Defines the depth of a fetch. Either empty, --depth=1, or --shallow-since=<>",
+        "reset_ref": """Reference to use for resetting the git repository.
+Either commit hash, tag or branch.""",
+        "fetch_ref": """Reference for fetching. Can be empty (HEAD), tag or branch.
+Can not be a commit hash, since typically it is forbidden by git servers.""",
+        "remote": "URL of the git repository to fetch from.",
+        "init_submodules": """If True, submodules update command will be called after fetching
+and resetting to the specified reference.""",
+    },
+)
+
+def git_repo(ctx, directory):
+    if ctx.attr.shallow_since:
+        if ctx.attr.tag:
+            fail("shallow_since not allowed if a tag is specified; --depth=1 will be used for tags")
+        if ctx.attr.branch:
+            fail("shallow_since not allowed if a branch is specified; --depth=1 will be used for branches")
+
+    shallow = "--depth=1"
+    if ctx.attr.shallow_since:
+        shallow = "--shallow-since=%s" % ctx.attr.shallow_since
+    if ctx.attr.commit:
+        shallow = ""
```

Comment:
This seems wrong. The reproducible case is that we specify a commit and a `shallow_since` value to allow shallow cloning. But your code seems to erase the value of `shallow` in this case.

Reply:
I will have another look tomorrow. (Of course the contract should stay).
Also, maybe you could point to a test for this specific combination?

NB this PR is mostly for passing my already-prepared changes to Yun, it's gonna be some work with this stuff before it gets submitted.

Reply:
The tests are now passing on Windows. @irengrig Can you address Klaus's comment or should I look into it? Is there anything else I should do before getting this merged?

Reply:
The current code works exactly the same way as the previous variant: https://github.com/bazelbuild/bazel/blob/220a8993c50eb5cd1cc2752fe46dd6760137432f/tools/build_defs/repo/git.bzl#L25

Before refactoring, if the commit was specified, and no shallow_since, the commit value was not used for shallow variable.
Please note that we can not call "fetch" with the commit number, only with tag/branch/HEAD (https://github.com/isaacs/github/issues/436), but maybe we could indeed use the --shallow_since attribute in this case.